fix(ivy): incorrectly adding class when falsy value is set in the presence of sanitizer

crisbeto opened the pull request:
Currently the styling instructions pass their sanitized value down to the function that decides whether to add or remove a style. The problem is that if a sanitizer is present, the value will be stringified which will throw off the falsy check inside the function. These changes fix the issue by passing in both the sanitized and unsanitized values, and using the raw value to do the check while assigning the sanitized one.

kara reviewed (approved):
LGTM, but I think @matsko will want to take a look

crisbeto committed:
fix(ivy): incorrectly adding class when falsy value is set in the presence of sanitizer

Currently the styling instructions pass their sanitized value down to the function that decides whether to add or remove a style. The problem is that if a sanitizer is present, the value will be stringified which will throw off the falsy check inside the function. These changes fix the issue by passing in both the sanitized and unsanitized values, and using the raw value to do the check while assigning the sanitized one.

crisbeto commented:
I went back-and-forth on a couple of solutions here, but I ended up reverting to my initial solution because it covered the most cases. We discussed it with @AndrewKushnir and we reached the conclusion that the way sanitization is set up at the moment is a little confusing. Setting the PR to blocked until we can discuss it on Monday.

AndrewKushnir commented:
Thanks for the fix @crisbeto!

FYI, we discussed these changes with @crisbeto offline. It looks non-intuitive that a presence of a single field that requires sanitization triggers sanitization for other style-related fields, like it happens with `ngClass`. I believe the processing logic in `ngClass` should be the same in case a template contains other sanitizable fields or not. I'm wondering if sanitization can be more granular and be applied only to the fields that actually need it?

// cc @matias @kara

kara commented:
@AndrewKushnir Yeah, I was thinking that we need to refactor styling sanitization, but I'm OK with doing that in a follow-up to make sure we get this in for RC (seems to be painful for users)

AndrewKushnir commented:
@kara thanks for the comment. Make sense, we can do refactoring in followup changes 👍

crisbeto commented:
Makes sense. In that case I'm unblocking the PR.

AndrewKushnir commented:
- [VE Presubmit](http://test/OCL:274480166:BASE:274480178:1571009552110:65a05bbd)
- [Ivy Presubmit](http://test/OCL:274480166:BASE:274480294:1571009635302:2ead171)

crisbeto commented:
Closing in favor of #33154.
